### DISCUSSION TITLE
Unusual behaviour clipping trajectories to with complex polygons

### DISCUSSION THREAD

**Discussion by EllieBowler** [anitagraser/movingpandas#235] [Q&A]
Hi, 

Really enjoying using movingpandas to explore some tracking data! I just had a question as for my task i need to look where my animal tracks intersect with an area of water. I've made a polygon of the water area and used the .clip function to try and get the segments within the bounds. However the water polygon is quite complex, with islands etc, i notice the clipped trajectories are not as i expected. 

I was able to replicate this using a toy example based on one of the example notebooks (1-tutorials/3-extracting-mover-positions.ipynb), when launched and run on binder. 

```
# make toy trajectory
df = pd.DataFrame([
  {'geometry':Point(0,2), 't':datetime(2018,1,1,12,0,0)},
  {'geometry':Point(6,2), 't':datetime(2018,1,1,12,6,0)},
  {'geometry':Point(6,6), 't':datetime(2018,1,1,12,10,0)},
  {'geometry':Point(9,9), 't':datetime(2018,1,1,12,15,0)},
]).set_index('t')
gdf = GeoDataFrame(df, crs=31256)
toy_traj = mpd.Trajectory(gdf, 1)

## make large polygon
xmin, xmax, ymin, ymax = 0, 5, 0, 5
polygon = Polygon([(xmin, ymin), (xmin, ymax), (xmax, ymax), (xmax, ymin), (xmin, ymin)])

## make smaller polygon
xmin2, xmax2, ymin2, ymax2 = 2, 3, 1, 3
smaller_polygon = Polygon([(xmin2, ymin2), (xmin2, ymax2), (xmax2, ymax2), (xmax2, ymin2), (xmin2, ymin2)])

## make extra smaller polygon
xmin3, xmax3, ymin3, ymax3 = 7, 8, 7, 8
extra_polygon = Polygon([(xmin3, ymin3), (xmin3, ymax3), (xmax3, ymax3), (xmax3, ymin3), (xmin3, ymin3)])

# combine into complex polygon shape
complex_poly = (polygon - smaller_polygon).union(extra_polygon)
complex_poly

complex_polygon_gdf = GeoDataFrame(pd.DataFrame([{'geometry':complex_poly, 'id':1}]), crs=31256)
intersections = toy_traj.clip(complex_poly)
ax = toy_traj.plot()
complex_polygon_gdf.plot(ax=ax, color='lightgray')
intersections.plot(ax=ax, color='red', linewidth=5, capstyle='round')
```
I get the output below. As you can see it does not pick up the segments in all the grey areas as i was expecting. I'm unsure if i'm misunderstanding how the clip function behaves, or if this is a bug with more complex shaped polygons?

![moving_pandas_clip](https://user-images.githubusercontent.com/39596309/182185174-8a281b9e-faee-4b5c-8bea-294e761a69bb.png)

Thanks for any suggestions!

**Comment by anitagraser** [comment]
Thank you for the reproducible example, Ellie. It's good to see that users are interested in the geometric functions of MovingPandas. 

I first thought that this may be a problem with MultiPolygons but that's not it:

![image](https://user-images.githubusercontent.com/590385/182443363-b1da5b2e-d072-491a-8763-dc6ac1baf9c1.png)

Instead, the issue is with the hole in the Polygon:

![image](https://user-images.githubusercontent.com/590385/182443523-552ee68c-418a-41dc-a9db-51b3f229ba31.png)

I've opened a corresponding ticket: https://github.com/anitagraser/movingpandas/issues/236

**Reply by EllieBowler** [reply]
Thanks for your quick reply and explanation Anita! I will keep an eye on the ticket for any updates. Do you have an approximate idea of roughly of how long a fix might take (i.e days/weeks)? Just so i can estimate a timeline for my analysis. I can try removing the holes in my water polygon in the meantime to get something approximate.

**Reply by anitagraser** [reply]
I expect to find some time in the second half of August. I'll update the ticket when I find a solution.
